### PR TITLE
WINC-528: Add metrics queries to display node graphs

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This commit ensures that we create a PrometheusRule object that
defines Windows specific rules. The queries defined utilze the
metrics generated from Windows exporter. Going forward, we need
to add required queries to the `windows.rules` group. The queries
added also result in displaying node graphs for Windows nodes in
OpenShift console.

### CODE REVIEW

File: test/e2e/metrics_test.go [hunk 1]
```diff
@@ -120,3 +120,13 @@ func checkTargetNodes(windowsEndpoints *core.Endpoints) error {
 
 	return nil
 }
+
+// testWindowsPrometheusRules tests if prometheus rules specific to Windows are defined by WMCO
+func testWindowsPrometheusRules(t *testing.T) {
```
Comment (mansikulkarni96) [inline test/e2e/metrics_test.go:125]: is this test being exercised somewhere?

File: deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+          record: node_network_transmit_bytes_total
+        - expr: |
+            windows_logical_disk_free_bytes
+          record: node_filesystem_free_bytes
+        - expr: |
+            windows_memory_available_bytes
+          record: node_memory_MemAvailable_bytes
+        - expr: |
+            windows_cs_physical_memory_bytes
+          record: node_memory_MemTotal_bytes
```
Comment (aravindhp) [inline deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml:38]: Please add a newline.

File: pkg/controller/windowsmachine/metrics/metrics.go [hunk 3]
```diff
@@ -52,7 +53,7 @@ type patchEndpoint struct {
 	Op string `json:"op"`
 	// path defines the location of the patch
 	Path string `json:"path"`
-	// value defines the data to be patched
+	//Value []v1.EndpointSubset `json:"value"`
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/metrics/metrics.go:56]: Space after `//`

File: pkg/controller/windowsmachine/metrics/metrics.go [hunk 4]
```diff
@@ -5,6 +5,7 @@ import (
 	"encoding/json"
 	"fmt"
 
+	monclientv1 "github.com/coreos/prometheus-operator/pkg/client/versioned/typed/monitoring/v1"
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/metrics/metrics.go:8]: Please name this `monclient`

File: deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml [hunk 5]
```diff
@@ -0,0 +1,38 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    prometheus: k8s
+    role: alert-rules
+  name: win-prometheus-k8s-rules
```
Comment (aravindhp) [inline deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml:7]: Please call this `windows-prometheus-...`

File: deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    - name: windows.rules
+      rules:
+        - expr: |
+            avg without (core, mode)(rate(windows_cpu_time_total{mode="idle"}[1m]))
+          record: instance:node_cpu_utilisation:rate1m
+        - expr: |
+            sum(rate(windows_cpu_time_total{mode="idle",mode!="iowait"}[3m])) BY (instance)
+          record: instance:node_cpu:rate:sum
+        - expr: |
+            windows_logical_disk_size_bytes
```
Comment (aravindhp) [inline deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml:19]: I assume `windows_logical_disk_size_bytes` is something `windows_exporter` is exposing.  

A logical disk needn't have all its space allocated. Are there other options for disk size that `windows_exporter` exposes?

File: deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            rate(windows_net_bytes_sent_total[1m])
+          record: node_network_transmit_bytes_total
+        - expr: |
+            windows_logical_disk_free_bytes
+          record: node_filesystem_free_bytes
+        - expr: |
+            windows_memory_available_bytes
+          record: node_memory_MemAvailable_bytes
+        - expr: |
+            windows_cs_physical_memory_bytes
```
Comment (aravindhp) [inline deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml:37]: What does `cs` stand for here?

File: pkg/controller/windowsmachine/metrics/metrics.go [hunk 8]
```diff
@@ -97,6 +98,10 @@ func Add(ctx context.Context, cfg *rest.Config, namespace string) error {
 
 		}
 	}
+	err = updateServiceMonitors(cfg, namespace)
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/metrics/metrics.go:101]: Is there a reason this cannot be done when creating the service monitors? PS: the question implies that a comment is required to explain why.

File: pkg/controller/windowsmachine/metrics/metrics.go [hunk 9]
```diff
@@ -229,3 +234,25 @@ func isEndpointsValid(nodes *v1.NodeList, endpoints *v1.Endpoints) bool {
 	}
 	return true
 }
+
+// updateServiceMonitors patches the metrics Service Monitor to include metrics path and replace action that replaces
+// node ip addressed returned by the windows exporter to node names. This is required to display node graphs on the
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/metrics/metrics.go:239]: Not sure what you are trying to say here, especially `replace action that replaces`. Also are you saying  `node IP addresses returned by windows_exporter with node names`?

File: test/e2e/metrics_test.go [hunk 10]
```diff
@@ -19,8 +19,12 @@ import (
 )
 
 func testMetrics(t *testing.T) {
+	err := framework.AddToFrameworkScheme(monitoringv1.AddToScheme, &monitoringv1.ServiceMonitor{})
+	require.NoError(t, err)
+
 	t.Run("Windows Exporter configuration validation", testWindowsExporter)
 	t.Run("Prometheus configuration validation", testPrometheus)
+	t.Run("Windows Prometheus rules validation", testWindowsPrometheusRules)
```
Comment (aravindhp) [inline test/e2e/metrics_test.go:27]: Is it possible to add a test that actually ensures we are seeing the expected metrics?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @VaishnaviHire. Please address my comments.

File: deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml (hunk 7 above)
Comment (VaishnaviHire) [reply deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml:37]: It is a metric defined by `windows-exporter`. https://github.com/prometheus-community/windows_exporter/blob/master/docs/collector.cs.md

File: pkg/controller/windowsmachine/metrics/metrics.go (hunk 9 above)
Comment (VaishnaviHire) [reply pkg/controller/windowsmachine/metrics/metrics.go:239]: I will update the comment.

Console graph queries require metrics endpoint target name to be `node name`, however `windows_exporter` returns `node IP`. We replace the target name by adding following to the `ServiceMonitor`

```
      path: "/metrics"
      relabelings:
      - action: replace
        regex: (.*)
        replacement: $1
        sourceLabels:
        - __meta_kubernetes_endpoint_address_target_name
        targetLabel: instance
```

Comment (aravindhp) [reply pkg/controller/windowsmachine/metrics/metrics.go:239]: Please add this level of detail to the comment.

File: test/e2e/metrics_test.go (hunk 10 above)
Comment (VaishnaviHire) [reply test/e2e/metrics_test.go:27]: [TestWindowsExporter](https://github.com/openshift/windows-machine-config-operator/blob/master/test/e2e/metrics_test.go#L55) tests if we can access the windows metrics provided by the `windows-exporter`. Do we want to test if those metrics are as expected ? One of the [tests](https://github.com/prometheus-community/windows_exporter/blob/master/tools/end-to-end-test.ps1) that `windows-exporter` do is to compare the output from `/metrics` endpoint to a given temp file.

Should testing metrics from `windows-exporter` be included in our repo as well ?

Comment (aravindhp) [reply test/e2e/metrics_test.go:27]: Let me state this in another way. You are adding a set of rules where you are mapping metrics exposed by `windows-exporter` to look like Linux metrics so that they work in the console. I was asking if there was a way to test if that is happening correctly.

File: deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml (hunk 6 above)
Comment (VaishnaviHire) [reply deploy/olm-catalog/windows-machine-config-operator/manifests/prometheusRule.yaml:19]: Did not see any other options for disk-size that can be used in `windows-exporter`
```
#HELP windows_logical_disk_free_bytes Free space in bytes (LogicalDisk.PercentFreeSpace)
# HELP windows_logical_disk_idle_seconds_total Seconds that the disk was idle (LogicalDisk.PercentIdleTime)
# HELP windows_logical_disk_read_bytes_total The number of bytes transferred from the disk during read operations (LogicalDisk.DiskReadBytesPerSec)
# HELP windows_logical_disk_read_latency_seconds_total Shows the average time, in seconds, of a read operation from the disk (LogicalDisk.AvgDiskSecPerRead)
# HELP windows_logical_disk_read_seconds_total Seconds that the disk was busy servicing read requests (LogicalDisk.PercentDiskReadTime)
# HELP windows_logical_disk_read_write_latency_seconds_total Shows the time, in seconds, of the average disk transfer (LogicalDisk.AvgDiskSecPerTransfer)
# HELP windows_logical_disk_reads_total The number of read operations on the disk (LogicalDisk.DiskReadsPerSec)
# HELP windows_logical_disk_requests_queued The number of requests queued to the disk (LogicalDisk.CurrentDiskQueueLength)
# HELP windows_logical_disk_size_bytes Total space in bytes (LogicalDisk.PercentFreeSpace_Base)
# HELP windows_logical_disk_split_ios_total The number of I/Os to the disk were split into multiple I/Os (LogicalDisk.SplitIOPerSec)
# HELP windows_logical_disk_write_bytes_total The number of bytes transferred to the disk during write operations (LogicalDisk.DiskWriteBytesPerSec)
# HELP windows_logical_disk_write_latency_seconds_total Shows the average time, in seconds, of a write operation to the disk (LogicalDisk.AvgDiskSecPerWrite)
# HELP windows_logical_disk_write_seconds_total Seconds that the disk was busy servicing write requests (LogicalDisk.PercentDiskWriteTime)
# HELP windows_logical_disk_writes_total The number of write operations on the disk (LogicalDisk.DiskWritesPerSec)
```

**Comment by VaishnaviHire** [comment]
New test log
```
--- PASS: TestWMCO (73.45s)
    --- PASS: TestWMCO/create (73.11s)
        --- PASS: TestWMCO/create/Creation (60.27s)
        --- PASS: TestWMCO/create/Metrics_validation (12.49s)
            --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation (12.49s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_instance:node_cpu_utilisation:rate1m (0.35s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_instance:node_cpu_utilisation:rate1m/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_instance:node_cpu_utilisation:rate1m/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_instance:node_cpu:rate:sum (0.18s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_instance:node_cpu:rate:sum/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_instance:node_cpu:rate:sum/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_size_bytes (0.07s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_size_bytes/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_size_bytes/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_avail_bytes (0.08s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_avail_bytes/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_avail_bytes/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_network_receive_bytes_total (0.15s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_network_receive_bytes_total/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_network_receive_bytes_total/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_network_transmit_bytes_total (0.10s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_network_transmit_bytes_total/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_network_transmit_bytes_total/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_free_bytes (0.08s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_free_bytes/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_filesystem_free_bytes/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_memory_MemAvailable_bytes (0.14s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_memory_MemAvailable_bytes/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_memory_MemAvailable_bytes/e2e-wmcow-sgmcs (0.00s)
                --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_memory_MemTotal_bytes (0.07s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_memory_MemTotal_bytes/e2e-wmcow-g7scg (0.00s)
                    --- PASS: TestWMCO/create/Metrics_validation/Windows_Prometheus_rules_validation/Query:_node_memory_MemTotal_bytes/e2e-wmcow-sgmcs (0.00s)
PASS
ok      github.com/openshift/windows-machine-config-operator/test/e2e   73.727s
```

File: pkg/controller/windowsmachine/metrics/metrics.go (hunk 8 above)
Comment (aravindhp) [reply pkg/controller/windowsmachine/metrics/metrics.go:101]: Did you forget to add the comment?

File: test/e2e/metrics_test.go [hunk 11]
```diff
@@ -120,3 +132,107 @@ func checkTargetNodes(windowsEndpoints *core.Endpoints) error {
...
+	// test if rules specific to windows exist
+	require.Equal(t, windowsRuleName, promRule.Spec.Groups[0].Name)
+
+	// get route to access Prometheus server instance in monitoring namespace
+	prometheusRoute := &route.Route{}
+	err = framework.Global.Client.Get(context.TODO(),
+		types.NamespacedName{Namespace: monitoringNamespace, Name: "prometheus-k8s"}, prometheusRoute)
+	require.NoError(t, err)
+
+	// define authorization token required to call Prometheus api
```
Comment (aravindhp) [inline test/e2e/metrics_test.go:182]: `API`

File: test/e2e/metrics_test.go [hunk 12]
```diff
@@ -120,3 +132,107 @@ func checkTargetNodes(windowsEndpoints *core.Endpoints) error {
...
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	client := &http.Client{Transport: tr}
+
+	// Following tests carry out api calls to Prometheus server with all the queries defined in the Windows PrometheusRule
+	// object. The tests check if the query results have instances corresponding to configured Windows Nodes.
+	// They also test if the metrics returned have a non-zero value.
+	for _, rules := range promRule.Spec.Groups {
+		if rules.Name == windowsRuleName {
```
Comment (aravindhp) [inline test/e2e/metrics_test.go:194]: One way to reduce indentation is to:
```suggestion
if rules.Name != windowsRuleName {
    continue
}
//rest of the code with less indentation
```

Comment (aravindhp) [reply test/e2e/metrics_test.go:194]: Now for whatever reason there is no Windows rules present, it will result in a false positive. So you need to check for that also.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/controller/windowsmachine/metrics/metrics.go (hunk 8 above)
Comment (VaishnaviHire) [reply pkg/controller/windowsmachine/metrics/metrics.go:101]: Added a detailed comment [here](https://github.com/openshift/windows-machine-config-operator/pull/234/files#diff-ccf7cfc3caf950d956bbe879bfbd3fd978235b1e00781499c93510d8e95ff994R238).

Comment (aravindhp) [reply pkg/controller/windowsmachine/metrics/metrics.go:101]: A method comment describes what the method does. You need a comment here at the call site as to why you need to call `updateServiceMonitors` soon after creating them.

File: test/e2e/metrics_test.go (hunk 12 above)
Comment (VaishnaviHire) [reply test/e2e/metrics_test.go:194]: We are testing for existence of `windows.rules` [here](https://github.com/openshift/windows-machine-config-operator/pull/234/commits/d541bb524846b3069608786d52617fbd7ff8a8eb#diff-41354c8d32cc5449ea0e5de5e5f28928103f9d2642b034f4b6e8474d487d656fR174)

**Review by aravindhp — APPROVED** [review]
Thanks for working on this, @VaishnaviHire 

/approve

**Review by mansikulkarni96 — COMMENTED** [review]
Thanks for working on this @VaishnaviHire 
LGTM except for a small nit.
